### ISSUE TITLE
python.Stock Price Prediction and Market Analysis System

### ISSUE THREAD

**Issue by AJAY-Git2004** [AJAY-Git2004/ajay_0E309042D772BFB990A46D32412DC5E5#1]
import yfinance as yf
import pandas as pd
import numpy as np

from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

def get_data(ticker):

    data = yf.Ticker(ticker).history(
        period="5y",
        interval="1d"
    )

    if data.empty:
        raise Exception("No data found for " + ticker)

    data = data[[
        "Open",
        "High",
        "Low",
        "Close",
        "Volume"
    ]].copy()

    return data

def create_features(data):

    data["SMA20"] = data["Close"].rolling(20).mean()

    data["SMA50"] = data["Close"].rolling(50).mean()

    data["EMA20"] = data["Close"].ewm(
        span=20,
        adjust=False
    ).mean()

    # RSI
    change = data["Close"].diff()

    gain = change.clip(lower=0)

    loss = -change.clip(upper=0)

    avg_gain = gain.rolling(14).mean()

    avg_loss = loss.rolling(14).mean()

    rs = avg_gain / avg_loss

    data["RSI"] = 100 - (100 / (1 + rs))

    # MACD
    ema12 = data["Close"].ewm(
        span=12,
        adjust=False
    ).mean()

    ema26 = data["Close"].ewm(
        span=26,
        adjust=False
    ).mean()

    data["MACD"] = ema12 - ema26

    # Target
    data["Target"] = data["Close"].shift(-1)

    data.dropna(inplace=True)

    return data

def train_model(data):

    features = [
        "Open",
        "High",
        "Low",
        "Close",
        "Volume",
        "SMA20",
        "SMA50",
        "EMA20",
        "RSI",
        "MACD"
    ]

    X = data[features]

    y = data["Target"]

    split = int(len(data) * 0.8)

    X_train = X.iloc[:split]

    X_test = X.iloc[split:]

    y_train = y.iloc[:split]

    y_test = y.iloc[split:]

    model = RandomForestRegressor(
        n_estimators=100,
        random_state=42
    )

    model.fit(
        X_train,
        y_train
    )

    predictions = model.predict(X_test)

    mae = mean_absolute_error(
        y_test,
        predictions
    )

    rmse = np.sqrt(
        mean_squared_error(
            y_test,
            predictions
        )
    )

    r2 = r2_score(
        y_test,
        predictions
    )

    return model, X_test, y_test, predictions, mae, rmse, r2

def predict(model, data):

    features = [
        "Open",
        "High",
        "Low",
        "Close",
        "Volume",
        "SMA20",
        "SMA50",
        "EMA20",
        "RSI",
        "MACD"
    ]

    latest = data[features].iloc[[-1]]

    prediction = model.predict(latest)[0]

    current = float(data["Close"].iloc[-1])

    change = ((prediction - current) / current) * 100

    if change > 1:
        signal = "BUY"
    elif change < -1:
        signal = "SELL"
    else:
        signal = "HOLD"

    return current, prediction, change, signal

# TEST PROGRAM
if __name__ == "__main__":

    ticker = "AAPL"

    print("Downloading data...")

    data = get_data(ticker)

    print("Data downloaded successfully!")

    print(data.tail())

    print("\nCreating features...")

    data = create_features(data)

    print("Features created successfully!")

    print("\nTraining model...")

    (
        model,
        X_test,
        y_test,
        predictions,
        mae,
        rmse,
        r2
    ) = train_model(data)

    print("Model trained successfully!")

    print("\nMODEL RESULTS")
    print("-------------------------")

    print("MAE:", round(mae, 4))

    print("RMSE:", round(rmse, 4))

    print("R2 Score:", round(r2, 4))

    current, predicted, change, signal = predict(
        model,
        data
    )

    print("\nPREDICTION")
    print("-------------------------")

    print(
        "Current Price:",
        round(current, 2)
    )

    print(
        "Predicted Price:",
        round(predicted, 2)
    )

    print(
        "Expected Change:",
        round(change, 2),
        "%"
    )

    print(
        "Signal:",
        signal
    )
